### ISSUE TITLE
client-s3: transformTo* Return Empty when S3 Object Has a Checksum

### ISSUE THREAD

**Issue by iheffernan** [aws/aws-sdk-js-v3#5107]
### Checkboxes for prior research

- [X] I've gone through [Developer Guide](https://docs.aws.amazon.com/sdk-for-javascript/v3/developer-guide) and [API reference](https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest)
- [X] I've checked [AWS Forums](https://forums.aws.amazon.com) and [StackOverflow](https://stackoverflow.com/questions/tagged/aws-sdk-js).
- [X] I've searched for [previous similar issues](https://github.com/aws/aws-sdk-js-v3/issues) and didn't find any solution.

### Describe the bug

When calling `GetObjectCommand` on an S3 object that was written with a `ChecksumAlgorithm`all of the `transformTo*` methods on the response `Body` return an empty string if `ChecksumMode: 'ENABLED'` was used.

In other words, if a Checksum exists on an S3 object, when calling `GetObject` I can either get a response that contains the Body payload or a response that contains the object's Checksum, but not both.

### SDK version number

@aws-sdk/client-s3@3.391.0

### Which JavaScript Runtime is this issue in?

Node.js

### Details of the browser/Node.js/ReactNative version

v18.17.1

### Reproduction Steps

```
const s3Client = new S3Client();
const bucket = 'my-bucket';
const key = 'test-key.txt';
const putCommand = new PutObjectCommand({
    Bucket: bucket,
    Key: key,
    Body: 'this is a test',
    ContentType: 'text/plain',
    ChecksumAlgorithm: 'SHA256'
});
const putResult = await s3Client.send(putCommand);

const getCommand = new GetObjectCommand({
    Bucket: bucket,
    Key: key,
    ChecksumMode: 'ENABLED'
});
const getResult = await s3Client.send(getCommand);
console.log(await getResult.Body?.transformToString('utf-8'));
```

### Observed Behavior

When the above code is run, the output is an empty string.

However, if you comment out or delete the `ChecksumMode: 'ENABLED'` in the `GetObjectCommand`, the output is correctly displayed as `this is a test`.

Alternatively, commenting out the `ChecksumAlgorithm: 'SHA256'` from the `PutObjectCommand` will also work, whether or not `ChecksumMode` is specified in the `GetObjectCommand`.

### Expected Behavior

I would expect the `Body.transformTo*` methods to correctly read the body whether or not a ChecksumAlgorithm was used when Put and ChecksumMode exists on the Get.

**Comment by yenfryherrerafeliz** [comment]
Hi @iheffernan, thanks for reporting this. I will add a needs-review label to this issue since I have confirmed the content is returned empty when ChecksumMode is set to ENABLED, so we can address this further. Basically, an internal exception happens when trying to calculate the hash for the body of the response.
Here is how the internal exception can be reproduced:

```js
import {GetObjectCommand, PutObjectCommand, S3Client} from "@aws-sdk/client-s3";
import crypto from "crypto";

const client = new S3Client({
    region: "us-east-2"
});
client.middlewareStack.addRelativeTo(
    (next) => async (args) => {
        const { response } = await next(args);
        const chunks = [];
        response.body.on('data', (chunk) => {
            chunks.push(chunk);
        })
        response.body.on('end', () => {
            console.log(Buffer.concat(chunks).toString('utf-8'))
        });
        return {
            response,
        };
    },
    {
        name: "nullFetchResponseBodyMiddleware",
        toMiddleware: "deserializerMiddleware",
        relation: "after",
        step: "deserialize",
        override: true,
    }
);
const body = "This is a test body";
const hash = crypto.createHash("sha256").update(body).digest('base64');
await client.send(new PutObjectCommand({
    Bucket: process.env.TEST_BUCKET,
    Key: process.env.TEST_KEY,
    Body: body,
    ChecksumSHA256: hash
}));
await client.send(new GetObjectCommand({
    Bucket: process.env.TEST_BUCKET,
    Key: process.env.TEST_KEY,
    ChecksumMode: 'ENABLED'
}));

```

Thanks!

**Comment by iheffernan** [comment]
Thanks, @yenfryherrerafeliz. I am able to reproduce the internal exception - `Unable to calculate hash for flowing readable stream` in `readableStreamHasher` - with the code above. Let me know if there's anything else you need to try to resolve this...

**Comment by yenfryherrerafeliz** [comment]
This issue seems to be introduced in version [3.383](https://github.com/aws/aws-sdk-js-v3/releases/tag/v3.383.0). The real issue behind this is that the body is being emptied when pipe to the hash calculator object [here](https://github.com/awslabs/smithy-typescript/blob/2d5a06af634c243bf2469566176dd17afedb1058/packages/hash-stream-node/src/readableStreamHasher.ts#L17), and this data is never returned back to the readable stream which represents the s3 object content. 

Thanks!

**Comment by trivikr** [comment]
I'm not able to reproduce issue when calling `getObject` for object which already exists. 
SDK versions tested: 3.382.0, 3.383.0 or 3.398.0

```js
// getObject.mjs
import { S3 } from "@aws-sdk/client-s3";

const region = "us-west-2"; // Replace with your S3 reqion, if required.
const Bucket = "aws-sdk-js-trivikr"; // Replace with Bucket in your AWS account.
const Key = "hello-world.txt"; // Replace with Key in your AWS account.
const ChecksumMode = "ENABLED";

const client = new S3({ region });
const response = await client.getObject({ Bucket, Key, ChecksumMode });
const objValue = await response.Body?.transformToString("utf-8");

console.log(objValue);
```

```console
$ node -v
v18.17.1

$ yarn --version
3.3.0

$ yarn add @aws-sdk/client-s3@3.382.0 --exact
...

$ yarn why @aws-sdk/client-s3
└─ test@workspace:.
   └─ @aws-sdk/client-s3@npm:3.382.0 (via npm:3.382.0)
   
$ node getObject.mjs 
helloworld

$  yarn add @aws-sdk/client-s3@3.383.0 --exact
...

$ yarn why @aws-sdk/client-s3                
└─ test@workspace:.
   └─ @aws-sdk/client-s3@npm:3.383.0 (via npm:3.383.0)

$ node getObject.mjs 
helloworld

$ yarn add @aws-sdk/client-s3@3.398.0 --exact
...

$ yarn why @aws-sdk/client-s3
└─ test@workspace:.
   └─ @aws-sdk/client-s3@npm:3.398.0 (via npm:3.398.0)
   
$ node getObject.mjs                         
helloworld
```

**Comment by trivikr** [comment]
Verified that the value returned is empty, only if the `ChecksumSHA256` was passed when calling PutObject

```js
import { S3 } from "@aws-sdk/client-s3";
import { createHash } from "crypto";

const region = "us-west-2"; // Replace with your S3 reqion, if required.
const Bucket = "aws-sdk-js-trivikr"; // Replace with Bucket in your AWS account.
const Key = "hello-world.txt"; // Replace with Key in your AWS account.
const Body = "helloworld"; // Replace with any body you're interested in..
const ChecksumMode = "ENABLED";
const ChecksumSHA256 = createHash("sha256").update(Body).digest("base64");

const client = new S3({ region });

// The getObject value is empty only when ChecksumSHA256 is passed, and client-s3 is >=3.383.0
// It returns non-empty body, if ChecksumSHA256 is not passed.
await client.putObject({ Bucket, Key, Body, ChecksumSHA256 });

const response = await client.getObject({ Bucket, Key, ChecksumMode });
const objValue = await response.Body?.transformToString("utf-8");

console.log(objValue);

```

**Comment by trivikr** [comment]
The use does not have to pass ChecksumSHA256 when calling putObject.
They can just set ChecksumAlgorithm to "SHA256", and the SDK will compute the checksum for them.

The issue is not reproducible with the following code in v3.383.0:
```js
import { S3 } from "@aws-sdk/client-s3";

const region = "us-west-2"; // Replace with your S3 reqion, if required.
const Bucket = "aws-sdk-js-trivikr"; // Replace with Bucket in your AWS account.
const Key = "hello-world.txt"; // Replace with Key in your AWS account.
const Body = "helloworld"; // Replace with any body you're interested in..
const ChecksumMode = "ENABLED";

const client = new S3({ region });

await client.putObject({ Bucket, Key, Body, ChecksumAlgorithm: "SHA256" });

const response = await client.getObject({ Bucket, Key, ChecksumMode });
const objValue = await response.Body?.transformToString("utf-8");

console.log(objValue);

```

**Comment by trivikr** [comment]
This issue is likely caused by something with Node.js event loop.

While testing, I sometimes am not able to reproduce the exact output mentioned in previous three comments.
I've hidden them as outdated, to not cause any confusion.

**Comment by iheffernan** [comment]
I think that the same fix that was applied in PR #5126 for another issue I had encountered (#5125) addresses whatever condition was causing this issue. I just grabbed the latest `client-s3` `v3.400.0` and both the response Body and SHA are accessible when a) a Checksum exists on the S3 object and b) the Object is retrieved with ChecksumMode = ENABLED.

Would be great if someone on the AWS side could confirm.
